spec.html

```diff
@@ -39276,69 +39038,84 @@ <h1>AsyncGeneratorRequest Records</h1>
         <emu-alg>
-          1. Assert: _generator_ is an AsyncGenerator instance.
+          1. Perform ? AsyncGeneratorValidate(_generator_, _generatorBrand_).
           1. Assert: _generator_.[[AsyncGeneratorState]] is *undefined*.
           1. Let _genContext_ be the running execution context.
           1. Set the Generator component of _genContext_ to _generator_.
           1. Set the code evaluation state of _genContext_ such that when evaluation is resumed for that execution context the following steps will be performed:
-            1. Let _result_ be the result of evaluating _generatorBody_.
+            1. If _generatorBody_ is a Parse Node, then
+              1. Let _result_ be the result of evaluating _generatorBody_.
+            1. Else,
+              1. Assert: _generatorBody_ is an Abstract Closure.
+              1. Let _result_ be _generatorBody_().
             1. Assert: If we return here, the async generator either threw an exception or performed either an implicit or explicit return.
             1. Remove _genContext_ from the execution context stack and restore the execution context that is at the top of the execution context stack as the running execution context.
             1. Set _generator_.[[AsyncGeneratorState]] to ~completed~.
             1. If _result_ is a normal completion, let _resultValue_ be *undefined*.
             1. Else,
               1. Let _resultValue_ be _result_.[[Value]].
               1. If _result_.[[Type]] is not ~return~, then
-                1. Return ! AsyncGeneratorReject(_generator_, _resultValue_).
-            1. Return ! AsyncGeneratorResolve(_generator_, _resultValue_, *true*).
+                1. Return ! AsyncGeneratorReject(_generator_, _resultValue_, _generatorBrand_).
+            1. Return ! AsyncGeneratorResolve(_generator_, _resultValue_, *true*, _generatorBrand_).
           1. Set _generator_.[[AsyncGeneratorContext]] to _genContext_.
           1. Set _generator_.[[AsyncGeneratorState]] to ~suspendedStart~.
           1. Set _generator_.[[AsyncGeneratorQueue]] to a new empty List.
           1. Return *undefined*.
         </emu-alg>
       </emu-clause>
 
+      <emu-clause id="sec-asyncgeneratorvalidate" aoid="AsyncGeneratorValidate">
+        <h1>AsyncGeneratorValidate ( _generator_, _generatorBrand_ )</h1>
+        <p>The abstract operation AsyncGeneratorValidate takes arguments _generator_ and _generatorBrand_. It performs the following steps when called:</p>
+        <emu-alg>
+          1. Perform ? RequireInternalSlot(_generator_, [[AsyncGeneratorContext]]).
+          1. Perform ? RequireInternalSlot(_generator_, [[AsyncGeneratorState]]).
+          1. Perform ? RequireInternalSlot(_generator_, [[AsyncGeneratorQueue]]).
+          1. If _generator_.[[GeneratorBrand]] is not equal to _generatorBrand_, throw a *TypeError* exception.
```

Comment:
```suggestion
          1. If _generator_.[[GeneratorBrand]] is not the same value as _generatorBrand_, throw a *TypeError* exception.
```

Reply:
```suggestion
          1. If ! SameValue(_generator_.[[GeneratorBrand]], _generatorBrand_) is *false*, throw a *TypeError* exception.
```

Reply:
That doesn't work, because `SameValue` only operates on ES language values, and `[[GeneratorBrand]]` might be `~empty~`, which is a spec value.